R/hmm_train.R

```diff
@@ -0,0 +1,126 @@
+#' other mlpack HMM tools.
+#'
+#' @param input_file File containing input observations.
+#' @param batch If true, input_file (and if passed, labels_file) are expected to
+#'   contain a list of files to use as input observation sequences (and label
+#'   sequences).  Default value "FALSE".
+#' @param gaussians Number of gaussians in each GMM (necessary when type is 'gmm'). 
+#'   Default value "0".
+#' @param input_model Pre-existing HMM model to initialize training with.
+#' @param labels_file Optional file of hidden states, used for labeled training. 
+#'   Default value "".
+#' @param seed Random seed.  If 0, 'std::time(NULL)' is used.  Default value "0".
+#' @param states Number of hidden states in HMM (necessary, unless model_file is
+#'   specified).  Default value "0".
+#' @param tolerance Tolerance of the Baum-Welch algorithm.  Default value "1e-05".
+#' @param type Type of HMM: discrete | gaussian | diag_gmm | gmm.  Default value
+#'   "gaussian".
+#' @param verbose Display informational messages and the full list of parameters and
+#'   timers at the end of execution.  Default value "FALSE".
+#'
+#' @return A list with several components:
+#' \item{output_model}{Output for trained HMM.}
+#'
+#' @details
+#' This program allows a Hidden Markov Model to be trained on labeled or
+#' unlabeled data.  It supports four types of HMMs: Discrete HMMs, Gaussian
+#' HMMs, GMM HMMs, or Diagonal GMM HMMs
+#' 
+#' Either one input sequence can be specified (with --input_file), or, a file
+#' containing files in which input sequences can be found (when --input_file and
+#' --batch are used together).  In addition, labels can be provided in the file
+#' specified by --labels_file, and if --batch is used, the file given to
+#' --labels_file should contain a list of files of labels corresponding to the
+#' sequences in the file given to --input_file.
+#' 
+#' The HMM is trained with the Baum-Welch algorithm if no labels are provided. 
+#' The tolerance of the Baum-Welch algorithm can be set with the --tolerance
+#' option.  By default, the transition matrix is randomly initialized and the
+#' emission distributions are initialized to fit the extent of the data.
+#' 
+#' Optionally, a pre-created HMM model can be used as a guess for the transition
+#' matrix and emission probabilities; this is specifiable with --model_file.
+#' @author
+#' MLPACK Developers
+#'
+#' @export
+
+hmm_train <- function(input_file,
+                      batch=FALSE,
+                      gaussians=NA,
+                      input_model=NA,
+                      labels_file=NA,
+                      seed=NA,
+                      states=NA,
+                      tolerance=NA,
+                      type=NA,
+                      verbose=FALSE) {
+  # Restore IO settings.
+  IO_RestoreSettings("Hidden Markov Model (HMM) Training")
+
+  # Process each input argument before calling mlpackMain().
+  IO_SetParamString("input_file", input_file)
```

Comment:
This is one of our "weird" bindings: because we don't have an automatic binding type for what the HMM training algorithm takes in, we instead just pass a filename to either a dataset or a file containing names of files that contain datasets.  I don't think there's anything to do here right now, but I at least wanted to point it out to @eddelbuettel and @coatless.  In the long term, this is one of a few bindings that we have to figure out a better representation for.